jetty-client/src/main/java/org/eclipse/jetty/client/util/AbstractRequestContent.java

```diff
@@ -0,0 +1,250 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.client.util;
+
+import java.io.EOFException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.thread.AutoLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractRequestContent implements Request.Content
+{
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractRequestContent.class);
+
+    private final AutoLock lock = new AutoLock();
+    private final String contentType;
+    private Subscription subscription;
+    private Throwable failure;
+
+    protected AbstractRequestContent(String contentType)
+    {
+        this.contentType = contentType;
+    }
+
+    @Override
+    public String getContentType()
+    {
+        return contentType;
+    }
+
+    @Override
+    public Subscription subscribe(Consumer consumer, boolean emitInitialContent)
+    {
+        Subscription oldSubscription;
+        Subscription newSubscription;
+        try (AutoLock ignored = lock.lock())
+        {
+            if (subscription != null && !isReproducible())
+                throw new IllegalStateException("Multiple subscriptions not supported on " + this);
+            oldSubscription = subscription;
+            newSubscription = subscription = newSubscription(consumer, emitInitialContent, failure);
+        }
+        if (oldSubscription != null)
+            oldSubscription.fail(new EOFException("Content replay"));
```

Comment:
why is this assuming that a content can't ber in simultaneous subscriptions?
Can't a reproducible content be sent to two different requests at the same time?

I don't think the abstract content should hold onto a subscription reference at all. it just needs logic like:
```java
    boolean oneTimeused;
    ...
    if (!isReproducible|())
    {
      if (oneTimeUsed)
          throw ISE(...);
      oneTimeUsed = true;
    }
    ...
```

Reply:
@gregw having multiple subscriptions for the same content is not a use case that we need implemented. If someone wants to do that, the API does not forbid it, but we don't need it.
It will be quite complicated to implement multiple subscriptions, as one will be slower than the other, and now you need a policy to decide whether the chunks should be provided independently, or wait for all the subscribers to demand, etc.

The fact that the subscription is overwritten in case of reproducible content is because the same content may be reproduced in a subsequent request (e.g. 302 or 401), so the subscriptions are sequential (not simultaneous), but only ever one is active.

Reply:
Why isn't it a use case we need to support?  I can see a web spider/crawler wanting to send the same content to many servers so why shouldn't we be able to wrap up a ByteBuffer as a content and send it  multiple times?

It is fine if some/many/most of our Contents cannot do this, as it would be complex for some.   But there is no reason to prevent it in the Abstract and no reason not to support it in the ByteBuffer case at the very least.